crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -344,49 +344,75 @@ impl CumulocityConverter {
-        match input.r#type {
-            EntityType::MainDevice => Ok(vec![]),
+        let reg_message = match input.r#type {
+            EntityType::MainDevice => None,
             EntityType::ChildDevice => {
                 let parent_xid: Option<&EntityExternalId> =
                     self.entity_cache.parent_external_id(entity_topic_id)?;
+                let display_name = display_name.unwrap_or(external_id.as_ref());
+                let display_type = display_type.unwrap_or("thin-edge.io-child");
 
                 let child_creation_message = child_device_creation_message(
                     external_id.as_ref(),
-                    display_name,
-                    display_type,
+                    Some(display_name),
+                    Some(display_type),
                     parent_xid.map(|xid| xid.as_ref()),
                     &self.device_name,
                     &self.config.bridge_config.c8y_prefix,
                 )
                 .context("Could not create device creation message")?;
-                Ok(vec![child_creation_message])
+                Some(child_creation_message)
             }
             EntityType::Service => {
                 let parent_xid = self.entity_cache.parent_external_id(entity_topic_id)?;
+                let display_name = display_name.unwrap_or_else(|| {
+                    entity_topic_id
+                        .default_service_name()
+                        .unwrap_or(external_id.as_ref())
+                });
+                let display_type = display_type.unwrap_or(&self.service_type);
 
                 let service_creation_message = service_creation_message(
                     external_id.as_ref(),
-                    display_name.unwrap_or_else(|| {
-                        entity_topic_id
-                            .default_service_name()
-                            .unwrap_or(external_id.as_ref())
-                    }),
-                    display_type.unwrap_or(&self.service_type),
+                    display_name,
+                    display_type,
                     "up",
                     parent_xid.map(|xid| xid.as_ref()),
                     &self.device_name,
                     &self.config.bridge_config.c8y_prefix,
                 )
                 .context("Could not create service creation message")?;
-                Ok(vec![service_creation_message])
+                Some(service_creation_message)
             }
+        };
+
+        if let Some(reg_message) = reg_message {
+            messages.push(reg_message);
         }
+
+        for (fragment_key, fragment_value) in input.other.iter() {
+            if fragment_key == "name" || fragment_key == "type" {
+                // Skip converting the name and type fields as they are already included in the registration message
+                continue;
+            }
+            let twin_messages = self.convert_twin_fragment(
```

Comment:
Isn't the mapper reacting *twice* on twin data? On the registration messages and on the twin messages derived by the agent from the registration messages?

Reply:
When the same twin data is delivered again, they are filtered out as "duplicate values" in the `inventory::try_convert_twin_fragment` function. Twin data is converted to inventory updates only if they really update the existing values. Else they are ignored.

Reply:
Resolved by a49d4cb